rules/DeadCode/Rector/FunctionLike/NarrowTooWideReturnTypeRector.php

```diff
@@ -0,0 +1,260 @@
+        private readonly SilentVoidResolver $silentVoidResolver,
+        private readonly PhpVersionProvider $phpVersionProvider,
+    ) {
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(
+            'Narrow too wide return type declarations if possible',
+            [
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+final class SomeClass
+{
+    public function foo(): string|int|\DateTime
+    {
+        if (rand(0, 1)) {
+            return 'text';
+        }
+
+        return 1000;
+    }
+}
+CODE_SAMPLE
+                    ,
+                    <<<'CODE_SAMPLE'
+final class SomeClass
+{
+    public function foo(): string|int
+    {
+        if (rand(0, 1)) {
+            return 'text';
+        }
+
+        return 1000;
+    }
+}
+CODE_SAMPLE
+                ),
+            ]
+        );
+    }
+
+    public function provideMinPhpVersion(): int
+    {
+        return PhpVersionFeature::UNION_TYPES;
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [ClassMethod::class, Function_::class, Closure::class, ArrowFunction::class];
+    }
+
+    /**
+     * @param ClassMethod|Function_|Closure|ArrowFunction $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        $scope = ScopeFetcher::fetch($node);
+
+        if ($this->shouldSkipNode($node, $scope)) {
+            return null;
+        }
+
+        $returnStatements = $node instanceof ArrowFunction
+            ? []
+            : $this->betterNodeFinder->findReturnsScoped($node);
+        $isAlwaysTerminating = ! $this->silentVoidResolver->hasSilentVoid($node);
+
+        if ($returnStatements === [] && ! $node instanceof ArrowFunction) {
+            if ($isAlwaysTerminating) {
+                if (! $this->phpVersionProvider->isAtLeastPhpVersion(PhpVersionFeature::NEVER_TYPE)) {
+                    return null;
+                }
+                $node->returnType = new Identifier('never');
+            } else {
+                $node->returnType = new Identifier('void');
+            }
+
+            return null;
+        }
+
+        $returnType = $node->returnType;
+
+        if (! $returnType instanceof UnionType) {
+            return null;
+        }
+
+        $actualReturnTypes = $this->collectActualReturnTypes($node, $returnStatements, $isAlwaysTerminating);
+        $newReturnType = $this->narrowUnionReturnType($returnType, $actualReturnTypes);
+
+        if ($newReturnType === null) {
+            return null;
+        }
+
+        $node->returnType = $newReturnType;
+
+        return $node;
+    }
+
+    private function shouldSkipNode(ClassMethod|Function_|Closure|ArrowFunction $node, Scope $scope): bool
+    {
+        $returnType = $node->returnType;
+
+        if (! $returnType instanceof UnionType && ! $returnType instanceof NullableType) {
+            return false;
+        }
+
```

Comment:
on ClassMethod, can add skip with `ClassMethodReturnTypeOverrideGuard::shouldSkipClassMethod()`, see example:

https://github.com/rectorphp/rector-src/blob/51dc24ca61e227e9f416886c69471bdaca0a7034/rules/TypeDeclaration/Rector/ClassMethod/BoolReturnTypeFromBooleanConstReturnsRector.php#L122-L123

Reply:
I'm not sure what you mean? I have tests for overridden methods...